File: src/fb.c
#include "fb.h"
#include "kprintf.h"

/* --- Limine framebuffer request ------------------------------------------- */
/* Minimal inline copy of the request/response structs (limine.h, v8/9 API). */

struct limine_framebuffer {
    void *address;
    uint64_t width;
    uint64_t height;
    uint64_t pitch;          /* bytes per scanline */
    uint16_t bpp;
    uint8_t  memory_model;
    uint8_t  red_mask_size,  red_mask_shift;
    uint8_t  green_mask_size, green_mask_shift;
    uint8_t  blue_mask_size,  blue_mask_shift;
    uint8_t  reserved[7];
};

struct limine_framebuffer_response {
    uint64_t revision;
    uint64_t framebuffer_count;
    struct limine_framebuffer **framebuffers;
};

struct limine_framebuffer_request {
    uint64_t id[4];
    uint64_t revision;
    struct limine_framebuffer_response *response;
};

__attribute__((used, section(".limine_requests")))
static volatile struct limine_framebuffer_request fb_request = {
    .id = { 0xc7b1dd30df4c8b88ULL, 0x0a82e883a194f07bULL,
            0x9d5827dcd881dd75ULL, 0xa3148604f6fab11bULL },
};

/* --- State ---------------------------------------------------------------- */

static volatile struct limine_framebuffer *fb;
static uint64_t fb_pitch;
static uint32_t fb_w, fb_h;
static uint8_t  fb_bpp;
static uint8_t  r_shift, g_shift, b_shift;
static int      fb_ready;

/* --- API ------------------------------------------------------------------ */

int fb_init(void) {
    if (!fb_request.response || fb_request.response->framebuffer_count < 1) {
        KLOG_ERR("no framebuffer from bootloader");
        return -1;
    }
    struct limine_framebuffer *f = fb_request.response->framebuffers[0];
    fb       = (volatile struct limine_framebuffer *)f;
    fb_pitch = f->pitch;
    fb_w     = (uint32_t)f->width;
    fb_h     = (uint32_t)f->height;
    fb_bpp   = (uint8_t)f->bpp;
    r_shift  = f->red_mask_shift;
    g_shift  = f->green_mask_shift;
    b_shift  = f->blue_mask_shift;
    fb_ready = 1;

    KLOG_INFO("fb: %ux%u %ubpp pitch=%lu addr=%p",
              fb_w, fb_h, fb_bpp, fb_pitch, f->address);
    KLOG_INFO("fb: rgb shifts r=%u g=%u b=%u", r_shift, g_shift, b_shift);
    return 0;
}

int fb_available(void) { return fb_ready; }
uint32_t fb_width(void)  { return fb_w; }
uint32_t fb_height(void) { return fb_h; }
uint8_t *fb_pixels(void) { return fb_ready ? (uint8_t *)fb->address : 0; }
uint32_t fb_pitch_bytes(void) { return (uint32_t)fb_pitch; }

fb_color_t fb_rgb(uint8_t r, uint8_t g, uint8_t b) {
    return ((fb_color_t)r << r_shift)
         | ((fb_color_t)g << g_shift)
         | ((fb_color_t)b << b_shift);
}

void fb_put_pixel(uint32_t x, uint32_t y, fb_color_t c) {
    if (!fb_ready || x >= fb_w || y >= fb_h)
        return;
    uint8_t *row = (uint8_t *)fb->address + (uint64_t)y * fb_pitch;
    if (fb_bpp == 32) {
        ((uint32_t *)row)[x] = (uint32_t)c;
    } else if (fb_bpp == 24) {
        uint8_t *p = row + (uint64_t)x * 3;
        p[0] = (uint8_t)(c & 0xFF);
        p[1] = (uint8_t)((c >> 8) & 0xFF);
        p[2] = (uint8_t)((c >> 16) & 0xFF);
    }
    /* other depths: unsupported in M2 */
}

void fb_fill_rect(uint32_t x, uint32_t y, uint32_t w, uint32_t h, fb_color_t c) {
    if (!fb_ready) return;
    if (x >= fb_w || y >= fb_h) return;
    if (x + w > fb_w) w = fb_w - x;
    if (y + h > fb_h) h = fb_h - y;

    for (uint32_t j = 0; j < h; j++) {
        uint8_t *row = (uint8_t *)fb->address + (uint64_t)(y + j) * fb_pitch;
        if (fb_bpp == 32) {
            uint32_t *p = (uint32_t *)row + x;
            for (uint32_t i = 0; i < w; i++)
                p[i] = (uint32_t)c;
        } else {
            for (uint32_t i = 0; i < w; i++)
                fb_put_pixel(x + i, y + j, c);
        }
    }
}

void fb_clear(fb_color_t c) {
    fb_fill_rect(0, 0, fb_w, fb_h, c);
}

void fb_test_pattern(void) {
    if (!fb_ready) return;

    /* background */
    fb_clear(fb_rgb(16, 16, 24));

    /* 8 classic colour bars, top half */
    static const uint8_t bars[8][3] = {
        {255, 255, 255}, {255, 255, 0}, {0, 255, 255}, {0, 255, 0},
        {255, 0, 255},   {255, 0, 0},   {0, 0, 255},   {0, 0, 0},
    };
    uint32_t bar_w = fb_w / 8;
    uint32_t bar_h = fb_h / 2;
    for (uint32_t i = 0; i < 8; i++)
        fb_fill_rect(i * bar_w, 0, bar_w, bar_h,
                     fb_rgb(bars[i][0], bars[i][1], bars[i][2]));

    /* horizontal gradient, bottom half (green = f(x), blue = f(y)) */
    for (uint32_t y = bar_h; y < fb_h; y += 2)
        for (uint32_t x = 0; x < fb_w; x += 64)
            fb_fill_rect(x, y, 64, 2,
                         fb_rgb(0,
                                (uint8_t)(x * 255 / fb_w),
                                (uint8_t)((y - bar_h) * 255 / (fb_h - bar_h))));

    /* white crosshairs to verify coordinate math at extremes */
    fb_fill_rect(fb_w / 2 - 1, 0, 2, fb_h, fb_rgb(255, 255, 255));
    fb_fill_rect(0, fb_h / 2 - 1, fb_w, 2, fb_rgb(255, 255, 255));
}

File: kernel-rs/src/fb.rs
//! Framebuffer over the Limine-provided linear buffer.

use crate::limine;

static mut FB_ADDR: *mut u8 = core::ptr::null_mut();
static mut FB_PITCH: u64 = 0;
static mut FB_W: u32 = 0;
static mut FB_H: u32 = 0;
static mut FB_BPP: u8 = 0;
static mut R_SHIFT: u8 = 0;
static mut G_SHIFT: u8 = 0;
static mut B_SHIFT: u8 = 0;
static mut FB_READY: bool = false;

pub type Color = u32;

pub fn init() -> i32 {
    let Some(r) = limine::framebuffer_response() else {
        crate::klog_err!("no framebuffer from bootloader");
        return -1;
    };
    if r.framebuffer_count < 1 {
        crate::klog_err!("no framebuffer from bootloader");
        return -1;
    }
    let f = unsafe { &*r.framebuffers.read() };
    unsafe {
        FB_ADDR = f.address;
        FB_PITCH = f.pitch;
        FB_W = f.width as u32;
        FB_H = f.height as u32;
        FB_BPP = f.bpp as u8;
        R_SHIFT = f.red_mask_shift;
        G_SHIFT = f.green_mask_shift;
        B_SHIFT = f.blue_mask_shift;
        FB_READY = true;
        crate::klog_info!(
            "fb: {}x{} {}bpp pitch={} addr={:#x}",
            FB_W,
            FB_H,
            FB_BPP,
            FB_PITCH,
            f.address as u64
        );
        crate::klog_info!(
            "fb: rgb shifts r={} g={} b={}",
            R_SHIFT,
            G_SHIFT,
            B_SHIFT
        );
    }
    0
}

pub fn available() -> bool {
    unsafe { FB_READY }
}

pub fn width() -> u32 {
    unsafe { FB_W }
}

pub fn height() -> u32 {
    unsafe { FB_H }
}

pub fn pixels() -> *mut u8 {
    if unsafe { FB_READY } {
        unsafe { FB_ADDR }
    } else {
        core::ptr::null_mut()
    }
}

pub fn pitch_bytes() -> u32 {
    unsafe { FB_PITCH as u32 }
}

pub fn rgb(r: u8, g: u8, b: u8) -> Color {
    unsafe { ((r as u32) << R_SHIFT) | ((g as u32) << G_SHIFT) | ((b as u32) << B_SHIFT) }
}

pub fn put_pixel(x: u32, y: u32, c: Color) {
    unsafe {
        if !FB_READY || x >= FB_W || y >= FB_H {
            return;
        }
        let row = FB_ADDR.add(y as usize * FB_PITCH as usize);
        if FB_BPP == 32 {
            *(row as *mut u32).add(x as usize) = c;
        } else if FB_BPP == 24 {
            let p = row.add(x as usize * 3);
            *p = (c & 0xFF) as u8;
            *p.add(1) = ((c >> 8) & 0xFF) as u8;
            *p.add(2) = ((c >> 16) & 0xFF) as u8;
        }
        // other depths: unsupported
    }
}

pub fn fill_rect(x: u32, y: u32, mut w: u32, mut h: u32, c: Color) {
    unsafe {
        if !FB_READY {
            return;
        }
        if x >= FB_W || y >= FB_H {
            return;
        }
        if x + w > FB_W {
            w = FB_W - x;
        }
        if y + h > FB_H {
            h = FB_H - y;
        }

        for j in 0..h {
            let row = FB_ADDR.add((y + j) as usize * FB_PITCH as usize);
            if FB_BPP == 32 {
                let p = (row as *mut u32).add(x as usize);
                for i in 0..w as usize {
                    *p.add(i) = c;
                }
            } else {
                for i in 0..w {
                    put_pixel(x + i, y + j, c);
                }
            }
        }
    }
}

pub fn clear(c: Color) {
    fill_rect(0, 0, width(), height(), c);
}

pub fn test_pattern() {
    if !available() {
        return;
    }

    // background
    clear(rgb(16, 16, 24));

    // 8 classic colour bars, top half
    const BARS: [[u8; 3]; 8] = [
        [255, 255, 255],
        [255, 255, 0],
        [0, 255, 255],
        [0, 255, 0],
        [255, 0, 255],
        [255, 0, 0],
        [0, 0, 255],
        [0, 0, 0],
    ];
    let (fb_w, fb_h) = (width(), height());
    let bar_w = fb_w / 8;
    let bar_h = fb_h / 2;
    for (i, b) in BARS.iter().enumerate() {
        fill_rect(i as u32 * bar_w, 0, bar_w, bar_h, rgb(b[0], b[1], b[2]));
    }

    // horizontal gradient, bottom half (green = f(x), blue = f(y))
    let mut y = bar_h;
    while y < fb_h {
        let mut x = 0;
        while x < fb_w {
            fill_rect(
                x,
                y,
                64,
                2,
                rgb(
                    0,
                    (x * 255 / fb_w) as u8,
                    ((y - bar_h) * 255 / (fb_h - bar_h)) as u8,
                ),
            );
            x += 64;
        }
        y += 2;
    }

    // white crosshairs to verify coordinate math at extremes
    fill_rect(fb_w / 2 - 1, 0, 2, fb_h, rgb(255, 255, 255));
    fill_rect(0, fb_h / 2 - 1, fb_w, 2, rgb(255, 255, 255));
}
